Confusing diagnostic when improperly using constant as type.

The following code...
```rust
#[allow(non_upper_case_globals)]
pub const Foo: &str = "";

mod submod {
    use super::Foo;
    fn function() {
        println!("{}", Foo::Bar);
    }
}
```

...yields the following error and warning:

```
error[E0433]: failed to resolve: use of undeclared type `Foo`
 --> src/lib.rs:7:24
  |
7 |         println!("{}", Foo::Bar);
  |                        ^^^ use of undeclared type `Foo`

warning: unused import: `super::Foo`
 --> src/lib.rs:5:9
  |
5 |     use super::Foo;
  |         ^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default
```

The warning says that `Foo` is not used but it clearly is, just not properly. This error is normally easier to spot when using the proper casing for constants, but that might not always be the case. Perhaps in this instance, we should look for other items with the exact same name and point out that the user is trying to use a constant as a type. 


<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"kwj2104"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

@henryboisdequin running the following will tell you where the error is being emitted:

```
% RUST_BACKTRACE=1 rustc +dev f5.rs -Ztreat-err-as-bug
error[E0433]: failed to resolve: use of undeclared type `Foo`
 --> f5.rs:7:24
  |
7 |         println!("{}", Foo::Bar);
  |                        ^^^ use of undeclared type `Foo`

thread 'rustc' panicked at 'aborting due to `-Z treat-err-as-bug=1`', compiler/rustc_errors/src/lib.rs:990:27
stack backtrace:
   0: std::panicking::begin_panic
   1: rustc_errors::HandlerInner::emit_diagnostic
   2: rustc_errors::Handler::emit_diagnostic
   3: rustc_errors::diagnostic_builder::DiagnosticBuilder::emit
   4: rustc_resolve::diagnostics::<impl rustc_resolve::Resolver>::report_error
   5: rustc_resolve::late::LateResolutionVisitor::report_error
   6: rustc_resolve::late::LateResolutionVisitor::smart_resolve_path_fragment
   7: rustc_resolve::late::LateResolutionVisitor::resolve_expr
   8: rustc_resolve::late::LateResolutionVisitor::resolve_expr
   9: rustc_ast::visit::walk_expr
  10: rustc_resolve::late::LateResolutionVisitor::resolve_expr
  11: rustc_ast::visit::walk_expr
  12: rustc_resolve::late::LateResolutionVisitor::resolve_expr
  13: rustc_resolve::late::LateResolutionVisitor::resolve_expr
  14: rustc_resolve::late::LateResolutionVisitor::resolve_expr
  15: rustc_resolve::late::LateResolutionVisitor::resolve_expr
  16: <rustc_resolve::late::LateResolutionVisitor as rustc_ast::visit::Visitor>::visit_block
  17: rustc_resolve::late::LateResolutionVisitor::resolve_expr
  18: <rustc_resolve::late::LateResolutionVisitor as rustc_ast::visit::Visitor>::visit_block
  19: rustc_resolve::late::LateResolutionVisitor::with_rib
  20: <rustc_resolve::late::LateResolutionVisitor as rustc_ast::visit::Visitor>::visit_fn
  21: rustc_ast::visit::walk_item
  22: rustc_resolve::late::LateResolutionVisitor::with_generic_param_rib
  23: <rustc_resolve::late::LateResolutionVisitor as rustc_ast::visit::Visitor>::visit_item
  24: rustc_ast::visit::walk_item
  25: rustc_resolve::late::LateResolutionVisitor::with_scope
  26: <rustc_resolve::late::LateResolutionVisitor as rustc_ast::visit::Visitor>::visit_item
  27: rustc_ast::visit::walk_crate
  28: rustc_resolve::late::<impl rustc_resolve::Resolver>::late_resolve_crate
  29: rustc_session::utils::<impl rustc_session::session::Session>::time
  30: rustc_resolve::Resolver::resolve_crate
  31: rustc_interface::passes::configure_and_expand_inner
  32: rustc_interface::passes::configure_and_expand::{{closure}}
  33: alloc::boxed::<impl core::ops::generator::Generator<R> for core::pin::Pin<alloc::boxed::Box<G,A>>>::resume
  34: rustc_data_structures::box_region::PinnedGenerator<I,A,R>::new
  35: rustc_interface::passes::BoxedResolver::new
  36: rustc_interface::passes::configure_and_expand
  37: rustc_interface::queries::Query<T>::compute
  38: rustc_interface::queries::Queries::expansion
  39: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter
  40: rustc_span::with_source_map
  41: rustc_interface::interface::create_compiler_and_run
  42: scoped_tls::ScopedKey<T>::set
  43: rustc_span::with_session_globals
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.51.0-dev running on x86_64-apple-darwin

note: compiler flags: -Z treat-err-as-bug

query stack during panic:
end of query stack
```

From there the place you might want to start looking at is `smart_resolve_path_fragment`, which is a bit hard to follow at the moment (I'm doing a slight clean up of that method at the moment), but should be reasonably straight forward to add another check for this case to that pile :)

--- a/compiler/rustc_resolve/src/late.rs
+++ b/compiler/rustc_resolve/src/late.rs
@@ -1596,7 +1596,7 @@ impl<'a: 'ast, 'b, 'ast> LateResolutionVisitor<'a, 'b, 'ast> {
                         .try_resolve_as_non_binding(pat_src, pat, bmode, ident, has_sub)
                         .unwrap_or_else(|| self.fresh_binding(ident, pat.id, pat_src, bindings));
                     self.r.record_partial_res(pat.id, PartialRes::new(res));
-                    self.r.record_local_span(pat.id, pat.span);
+                    self.r.record_pat_span(pat.id, pat.span);
                 }
                 PatKind::TupleStruct(ref path, ref sub_patterns) => {
                     self.smart_resolve_path(

--- a/compiler/rustc_resolve/src/late.rs
+++ b/compiler/rustc_resolve/src/late.rs
@@ -1031,7 +1031,6 @@ impl<'a: 'ast, 'b, 'ast> LateResolutionVisitor<'a, 'b, 'ast> {
             }
 
             ItemKind::Static(ref ty, _, ref expr) | ItemKind::Const(_, ref ty, ref expr) => {
-                debug!("resolve_item ItemKind::Const");
                 self.with_item_rib(HasGenericParams::No, |this| {
                     this.visit_ty(ty);
                     if let Some(expr) = expr {
@@ -1597,6 +1596,7 @@ impl<'a: 'ast, 'b, 'ast> LateResolutionVisitor<'a, 'b, 'ast> {
                         .try_resolve_as_non_binding(pat_src, pat, bmode, ident, has_sub)
                         .unwrap_or_else(|| self.fresh_binding(ident, pat.id, pat_src, bindings));
                     self.r.record_partial_res(pat.id, PartialRes::new(res));
+                    self.r.record_local_span(pat.id, pat.span);
                 }
                 PatKind::TupleStruct(ref path, ref sub_patterns) => {
                     self.smart_resolve_path(

--- a/compiler/rustc_resolve/src/lib.rs
+++ b/compiler/rustc_resolve/src/lib.rs
@@ -884,9 +884,9 @@ pub struct Resolver<'a> {
     /// "self-confirming" import resolutions during import validation.
     unusable_binding: Option<&'a NameBinding<'a>>,
 
-    // Spans for local variables found during resolution
-    // Used for suggestions during error reporting
-    local_span_map: NodeMap<Span>,
+    // Spans for local variables found during pattern resolution.
+    // Used for suggestions during error reporting.
+    pat_span_map: NodeMap<Span>,
 
     /// Resolutions for nodes that have a single resolution.
     partial_res_map: NodeMap<PartialRes>,
@@ -1266,7 +1266,7 @@ impl<'a> Resolver<'a> {
             last_import_segment: false,
             unusable_binding: None,
 
-            local_span_map: Default::default(),
+            pat_span_map: Default::default(),
             partial_res_map: Default::default(),
             import_res_map: Default::default(),
             label_res_map: Default::default(),
@@ -1884,6 +1884,7 @@ impl<'a> Resolver<'a> {
                     ribs,
                 )));
             }
+
             module = match ribs[i].kind {
                 ModuleRibKind(module) => module,
                 MacroDefinition(def) if def == self.macro_def(ident.span.ctxt()) => {
@@ -1894,6 +1895,7 @@ impl<'a> Resolver<'a> {
                 }
                 _ => continue,
             };
+
             match module.kind {
                 ModuleKind::Block(..) => {} // We can see through blocks
                 _ => break,
@@ -1912,19 +1914,16 @@ impl<'a> Resolver<'a> {
                 return Some(LexicalScopeBinding::Item(binding));
             }
         }
-        let returned_item = self
-            .early_resolve_ident_in_lexical_scope(
-                orig_ident,
-                ScopeSet::Late(ns, module, record_used_id),
-                parent_scope,
-                record_used,
-                record_used,
-                path_span,
-            )
-            .ok()
-            .map(LexicalScopeBinding::Item);
-
-        returned_item
+        self.early_resolve_ident_in_lexical_scope(
+            orig_ident,
+            ScopeSet::Late(ns, module, record_used_id),
+            parent_scope,
+            record_used,
+            record_used,
+            path_span,
+        )
+        .ok()
+        .map(LexicalScopeBinding::Item)
     }
 
     fn hygienic_lexical_parent(
@@ -2391,11 +2390,9 @@ impl<'a> Resolver<'a> {
                             .next()
                             .map_or(false, |c| c.is_ascii_uppercase())
                         {
-                            // Add check case for similarly named item in alternative namespace
-                            let mut suggestion = None;
-
-                            if ribs.is_some() {
-                                if let Some(res) = self.resolve_ident_in_lexical_scope(
+                            // Check whether the name refers to an item in the value namespace.
+                            let suggestion = if ribs.is_some() {
+                                let match_span = match self.resolve_ident_in_lexical_scope(
                                     ident,
                                     ValueNS,
                                     parent_scope,
@@ -2403,26 +2400,47 @@ impl<'a> Resolver<'a> {
                                     path_span,
                                     &ribs.unwrap()[ValueNS],
                                 ) {
-                                    let mut match_span: Option<Span> = None;
-                                    match res {
-                                        LexicalScopeBinding::Res(Res::Local(id)) => {
-                                            match_span =
-                                                Some(*self.local_span_map.get(&id).unwrap());
-                                        }
-                                        LexicalScopeBinding::Item(name_binding) => {
-                                            match_span = Some(name_binding.span);
-                                        }
-                                        _ => (),
-                                    };
-                                    if let Some(span) = match_span {
-                                        suggestion = Some((
-                                            vec![(span, String::from(""))],
-                                            format!("{} is defined here, but is not a type", ident),
-                                            Applicability::MaybeIncorrect,
-                                        ));
+                                    // Name matches a local variable. For example:
+                                    // ```
+                                    // fn f() {
+                                    //     let Foo: &str = "";
+                                    //     println!("{}", Foo::Bar); // Name refers to local
+                                    //                               // variable `Foo`.
+                                    // }
+                                    // ```
+                                    Some(LexicalScopeBinding::Res(Res::Local(id))) => {
+                                        Some(*self.pat_span_map.get(&id).unwrap())
                                     }
+
+                                    // Name matches item from a local name binding
+                                    // created by `use` declaration. For example:
+                                    // ```
+                                    // pub Foo: &str = "";
+                                    //
+                                    // mod submod {
+                                    //     use super::Foo;
+                                    //     println!("{}", Foo::Bar); // Name refers to local
+                                    //                               // binding `Foo`.
+                                    // }
+                                    // ```
+                                    Some(LexicalScopeBinding::Item(name_binding)) => {
+                                        Some(name_binding.span)
+                                    }
+                                    _ => None,
+                                };
+
+                                if let Some(span) = match_span {
+                                    Some((
+                                        vec![(span, String::from(""))],
+                                        format!("`{}` is defined here, but is not a type", ident),
+                                        Applicability::MaybeIncorrect,
+                                    ))
+                                } else {
+                                    None
                                 }
-                            }
+                            } else {
+                                None
+                            };
 
                             (format!("use of undeclared type `{}`", ident), suggestion)
                         } else {
@@ -2835,9 +2853,9 @@ impl<'a> Resolver<'a> {
         }
     }
 
-    fn record_local_span(&mut self, node: NodeId, span: Span) {
-        debug!("(recording local) recording {:?} for {:?}", node, span);
-        self.local_span_map.insert(node, span);
+    fn record_pat_span(&mut self, node: NodeId, span: Span) {
+        debug!("(recording pat) recording {:?} for {:?}", node, span);
+        self.pat_span_map.insert(node, span);
     }
 
     fn is_accessible_from(&self, vis: ty::Visibility, module: Module<'a>) -> bool {

--- a/compiler/rustc_resolve/src/lib.rs
+++ b/compiler/rustc_resolve/src/lib.rs
@@ -884,6 +884,10 @@ pub struct Resolver<'a> {
     /// "self-confirming" import resolutions during import validation.
     unusable_binding: Option<&'a NameBinding<'a>>,
 
+    // Spans for local variables found during resolution
+    // Used for suggestions during error reporting
+    local_span_map: NodeMap<Span>,
+
     /// Resolutions for nodes that have a single resolution.
     partial_res_map: NodeMap<PartialRes>,
     /// Resolutions for import nodes, which have multiple resolutions in different namespaces.
@@ -1262,6 +1266,7 @@ impl<'a> Resolver<'a> {
             last_import_segment: false,
             unusable_binding: None,
 
+            local_span_map: Default::default(),
             partial_res_map: Default::default(),
             import_res_map: Default::default(),
             label_res_map: Default::default(),
@@ -1879,7 +1884,6 @@ impl<'a> Resolver<'a> {
                     ribs,
                 )));
             }
-
             module = match ribs[i].kind {
                 ModuleRibKind(module) => module,
                 MacroDefinition(def) if def == self.macro_def(ident.span.ctxt()) => {
@@ -1890,7 +1894,6 @@ impl<'a> Resolver<'a> {
                 }
                 _ => continue,
             };
-
             match module.kind {
                 ModuleKind::Block(..) => {} // We can see through blocks
                 _ => break,
@@ -1909,17 +1912,19 @@ impl<'a> Resolver<'a> {
                 return Some(LexicalScopeBinding::Item(binding));
             }
         }
+        let returned_item = self
+            .early_resolve_ident_in_lexical_scope(
+                orig_ident,
+                ScopeSet::Late(ns, module, record_used_id),
+                parent_scope,
+                record_used,
+                record_used,
+                path_span,
+            )
+            .ok()
+            .map(LexicalScopeBinding::Item);
 
-        self.early_resolve_ident_in_lexical_scope(
-            orig_ident,
-            ScopeSet::Late(ns, module, record_used_id),
-            parent_scope,
-            record_used,
-            record_used,
-            path_span,
-        )
-        .ok()
-        .map(LexicalScopeBinding::Item)
+        returned_item
     }
 
     fn hygienic_lexical_parent(
@@ -2386,7 +2391,40 @@ impl<'a> Resolver<'a> {
                             .next()
                             .map_or(false, |c| c.is_ascii_uppercase())
                         {
-                            (format!("use of undeclared type `{}`", ident), None)
+                            // Add check case for similarly named item in alternative namespace
+                            let mut suggestion = None;
+
+                            if ribs.is_some() {
+                                if let Some(res) = self.resolve_ident_in_lexical_scope(
+                                    ident,
+                                    ValueNS,
+                                    parent_scope,
+                                    None,
+                                    path_span,
+                                    &ribs.unwrap()[ValueNS],
+                                ) {
+                                    let mut match_span: Option<Span> = None;
+                                    match res {
+                                        LexicalScopeBinding::Res(Res::Local(id)) => {
+                                            match_span =
+                                                Some(*self.local_span_map.get(&id).unwrap());
+                                        }
+                                        LexicalScopeBinding::Item(name_binding) => {
+                                            match_span = Some(name_binding.span);
+                                        }
+                                        _ => (),
+                                    };
+                                    if let Some(span) = match_span {
+                                        suggestion = Some((
+                                            vec![(span, String::from(""))],
+                                            format!("{} is defined here, but is not a type", ident),
+                                            Applicability::MaybeIncorrect,
+                                        ));
+                                    }
+                                }
+                            }
+
+                            (format!("use of undeclared type `{}`", ident), suggestion)
                         } else {
                             (format!("use of undeclared crate or module `{}`", ident), None)
                         }
@@ -2795,6 +2833,11 @@ impl<'a> Resolver<'a> {
         if let Some(prev_res) = self.partial_res_map.insert(node_id, resolution) {
             panic!("path resolved multiple times ({:?} before, {:?} now)", prev_res, resolution);
         }
+    }
+
+    fn record_local_span(&mut self, node: NodeId, span: Span) {
+        debug!("(recording local) recording {:?} for {:?}", node, span);
+        self.local_span_map.insert(node, span);
     }
 
     fn is_accessible_from(&self, vis: ty::Visibility, module: Module<'a>) -> bool {

--- a/src/test/ui/resolve/issue-81508.rs
+++ b/src/test/ui/resolve/issue-81508.rs
@@ -1,0 +1,22 @@
+// Confusing diagnostic when using variable as a type:
+//
+// Previous warnings indicate Foo is not used, when in fact it is
+// used improperly as a variable or constant. New warning points
+// out user may be trying to use variable as a type. Test demonstrates
+// cases for both local variable and const.
+
+fn main() {
+    let Baz: &str = "";
+
+    println!("{}", Baz::Bar); //~ ERROR: failed to resolve: use of undeclared type `Baz`
+}
+
+#[allow(non_upper_case_globals)]
+pub const Foo: &str = "";
+
+mod submod {
+    use super::Foo;
+    fn function() {
+        println!("{}", Foo::Bar); //~ ERROR: failed to resolve: use of undeclared type `Foo`
+    }
+}

--- a/src/test/ui/resolve/issue-81508.stderr
+++ b/src/test/ui/resolve/issue-81508.stderr
@@ -2,7 +2,7 @@ error[E0433]: failed to resolve: use of undeclared type `Baz`
   --> $DIR/issue-81508.rs:11:20
    |
 LL |     let Baz: &str = "";
-   |         --- help: Baz is defined here, but is not a type
+   |         --- help: `Baz` is defined here, but is not a type
 LL | 
 LL |     println!("{}", Baz::Bar);
    |                    ^^^ use of undeclared type `Baz`
@@ -11,7 +11,7 @@ error[E0433]: failed to resolve: use of undeclared type `Foo`
   --> $DIR/issue-81508.rs:20:24
    |
 LL |     use super::Foo;
-   |         ---------- help: Foo is defined here, but is not a type
+   |         ---------- help: `Foo` is defined here, but is not a type
 LL |     fn function() {
 LL |         println!("{}", Foo::Bar);
    |                        ^^^ use of undeclared type `Foo`

--- a/src/test/ui/resolve/issue-81508.stderr
+++ b/src/test/ui/resolve/issue-81508.stderr
@@ -1,0 +1,21 @@
+error[E0433]: failed to resolve: use of undeclared type `Baz`
+  --> $DIR/issue-81508.rs:11:20
+   |
+LL |     let Baz: &str = "";
+   |         --- help: Baz is defined here, but is not a type
+LL | 
+LL |     println!("{}", Baz::Bar);
+   |                    ^^^ use of undeclared type `Baz`
+
+error[E0433]: failed to resolve: use of undeclared type `Foo`
+  --> $DIR/issue-81508.rs:20:24
+   |
+LL |     use super::Foo;
+   |         ---------- help: Foo is defined here, but is not a type
+LL |     fn function() {
+LL |         println!("{}", Foo::Bar);
+   |                        ^^^ use of undeclared type `Foo`
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0433`.

